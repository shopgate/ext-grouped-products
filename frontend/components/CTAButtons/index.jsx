import React from 'react';
import PropTypes from 'prop-types';
import FavoritesButton from '@shopgate/pwa-ui-shared/FavoritesButton';
import AddToCartPicker from '../AddToCartPicker';
import { ADD_TO_CART_BUTTON_TYPE_DEFAULT } from '../../constants';
import styles from './style';
import connect from './connector';

/**
 * Renders CTA buttons for product page (add to cart + toggle favorites).
 *
 * @param {Object} props Props.
 * @returns {JSX}
 * @constructor
 */
const CTAButtons = (props) => {
  const addToCartButtonProps = {
    ...props.addToCartButtonProps,
    buttonSize: styles.cartButtonSize,
    iconSize: styles.iconSize,
    type: ADD_TO_CART_BUTTON_TYPE_DEFAULT,
  };

  const { id, stock } = props.product;

  return (
    <div className={styles.buttons}>
      <FavoritesButton
        active={props.isFavorite}
        productId={id}
        className={styles.favButton}
        rippleClassName={styles.ripple}
      />

      { props.isAddToCartButtonVisible && (
        <AddToCartPicker
          buttonProps={addToCartButtonProps}
          handleAddToCart={props.handleAddToCart}
          productId={id}
          stock={stock}
        />
      )}
    </div>
  );
};

CTAButtons.propTypes = {
  handleAddToCart: PropTypes.func.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  addToCartButtonProps: PropTypes.shape(),
  isAddToCartButtonVisible: PropTypes.bool,
  product: PropTypes.shape(),
};

CTAButtons.defaultProps = {
  addToCartButtonProps: {},
  isAddToCartButtonVisible: true,
  product: {},
};

export default connect(CTAButtons);
