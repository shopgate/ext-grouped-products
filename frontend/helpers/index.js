import { generateResultHash } from '@shopgate/pwa-common/helpers/redux';
import appConfig from '@shopgate/pwa-common/helpers/config';
import { SHOPGATE_CATALOG_GET_PRODUCT_CHILDREN, THEME_GMD } from '../constants';
import { maxQuantityPickerEntries as maxEntries } from '../config';

/**
 * Generates a hash to read and write with the productsByHash store.
 * @param {string} productId A product id.
 * @return {string}
 */
export const generateHash = productId => generateResultHash({
  pipeline: SHOPGATE_CATALOG_GET_PRODUCT_CHILDREN,
  productId,
});

/**
 * Creates an item list for the AddToCartPicker.
 * @param {Object} stock The stock property of a product.
 * @return {Array}
 */
export const createPickerItems = (stock) => {
  if (!stock) {
    return [];
  }

  const {
    ignoreQuantity,
    quantity,
  } = stock;

  const items = [];

  for (let i = 0; i <= maxEntries; i += 1) {
    if (ignoreQuantity) {
      if (items.length === maxEntries) {
        // When the quantity can be ignored the maximum number of entries can be created.
        break;
      }
    } else if (items.length === maxEntries || i >= quantity) {
      // When the quantity can't be ignored the maxium number of entries is limited.
      break;
    }

    const value = i + 1;

    items.push({
      label: `${value}`,
      value,
    });
  }

  return items;
};

/**
 * Checks if the current theme is the GMD theme.
 * @return {boolean}
 */
export const isGmdTheme = () => process.env.THEME.includes(THEME_GMD);

/**
 * Tells if the extension shall render flat buttons.
 * @return {boolean}
 */
export const renderFlatButtons = () => !isGmdTheme();

/**
 * Check if the favorite list feature is activated.
 * @return {boolean}
 */
export const hasFavorites = () => appConfig.hasFavorites;
