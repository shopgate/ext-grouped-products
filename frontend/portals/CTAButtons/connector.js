import { connect } from 'react-redux';
import {
  getCurrentProduct,
  isProductOrderable,
} from '@shopgate/pwa-common-commerce/product/selectors/product';
import { isCurrentProductOnFavoriteList } from '@shopgate/pwa-common-commerce/favorites/selectors';
import addCurrentProductToCart from '@shopgate/pwa-common-commerce/cart/actions/addCurrentProductToCart';
import setProductQuantity from '@shopgate/pwa-common-commerce/product/action-creators/setProductQuantity';
import {
  isProductPageLoading,
  isProductPageOrderable,
} from '@shopgate/pwa-common-commerce/product/selectors/page';
import { isMainAddToCartButtonVisible } from '../../selectors/index';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @return {Object} The extended component props.
 */
const mapStateToProps = state => ({
  addToCartButtonProps: {
    isLoading: isProductPageLoading(state),
    isOrderable: isProductPageOrderable(state),
    isDisabled: !isProductOrderable(state),
  },
  isAddToCartButtonVisible: isMainAddToCartButtonVisible(state),
  isFavorite: isCurrentProductOnFavoriteList(state),
  product: getCurrentProduct(state) || {},
});

/**
 * Connects the dispatch function to a callable function in the props.
 * @param {Function} dispatch The redux dispatch function.
 * @return {Object} The extended component props.
 */
const mapDispatchToProps = dispatch => ({
  handleAddToCart: (quantity) => {
    dispatch(setProductQuantity(quantity));
    dispatch(addCurrentProductToCart());
  },
});

export default connect(mapStateToProps, mapDispatchToProps);
