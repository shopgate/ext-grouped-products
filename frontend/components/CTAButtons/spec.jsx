import React from 'react';
import { shallow } from 'enzyme';
import { renderFlatButtons } from '../../helpers/index';
import CTAButtons from './index';

jest.mock('../../components/AddToCartPicker/index', () => props => <div {...props}>Hello world</div>);
jest.mock('../../helpers/index', () => ({
  renderFlatButtons: jest.fn(),
}));
const handleAddToCart = jest.fn();
describe('<CTAButtons />', () => {
  it('should Render both buttons when isAddToCartButtonVisible is true', () => {
    const wrapper = shallow(<CTAButtons
      handleAddToCart={handleAddToCart}
      isFavorite
      isAddToCartButtonVisible
      renderFlatButtons={renderFlatButtons}
    />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should only render FavoritesButton when isAddToCartButtonVisible is not true', () => {
    const wrapper = shallow(<CTAButtons
      handleAddToCart={handleAddToCart}
      isFavorite
      isAddToCartButtonVisible={false}
      renderFlatButtons={renderFlatButtons}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
