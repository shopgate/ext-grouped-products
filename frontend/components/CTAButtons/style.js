import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { variables } = themeConfig;

const cartButtonSize = 56;
const iconSize = 24;

const buttons = css({
  position: 'absolute',
  right: variables.gap.big,
  top: -30,
  height: 60,
  display: 'flex',
  justifyContent: 'flex-end',
  alignItems: 'center',
  '> :last-child': {
    marginLeft: variables.gap.big,
  },
}).toString();

const favButton = css({
  zIndex: 1, // Prevents the icons to be visible outside of the circle
  fontSize: iconSize,
}).toString();

const ripple = css({
  padding: 8,
}).toString();

export default {
  buttons,
  cartButtonSize,
  favButton,
  iconSize,
  ripple,
};
