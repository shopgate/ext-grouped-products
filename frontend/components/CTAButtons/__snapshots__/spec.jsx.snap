// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CTAButtons /> should Render both buttons when isAddToCartButtonVisible is true 1`] = `
<div
  className="css-1ldzh7h"
>
  <Connect(FavoritesButton)
    active={true}
    className="css-h7rvbh"
    rippleClassName="css-1tdptiz"
  />
  <AddToCartPicker
    buttonProps={
      Object {
        "buttonSize": 56,
        "iconSize": 24,
        "type": "default",
      }
    }
    clickDelay={150}
    handleAddToCart={[MockFunction]}
    productId=""
    stock={null}
  />
</div>
`;

exports[`<CTAButtons /> should only render FavoritesButton when isAddToCartButtonVisible is not true 1`] = `
<div
  className="css-1ldzh7h"
>
  <Connect(FavoritesButton)
    active={true}
    className="css-h7rvbh"
    rippleClassName="css-1tdptiz"
  />
</div>
`;
